runtime/tchannel_client.go

```diff
@@ -207,10 +232,33 @@ func (c *TChannelClient) call(
 			sc = c.scAlt
 		}
 
-		call.call, cerr = sc.BeginCall(ctx, call.serviceMethod, &tchannel.CallOptions{
-			Format:       tchannel.Thrift,
-			RequestState: rs,
-		})
+		if hostport != "" {
+			c.selectedPeers[hostport] = struct{}{}
+			p := sc.Peers().Add(hostport)
+			call.call, cerr = p.BeginCall(ctx, sc.ServiceName(), call.serviceMethod, &tchannel.CallOptions{
+				Format:       tchannel.Thrift,
+				RequestState: rs,
+			})
+
+			defer func() {
+				sc.Peers().Remove(hostport)
+				delete(c.selectedPeers, hostport)
```

Comment:
I don't see how this is thread safe. Can you please add a test that proves it using the race detector? Specifically, what if a request to a client is happening at the same time as this deferred function? Won't the `delete(...)` conflict with the assignment `c.selectedPeers[hostport] = ...`?

Reply:
Yes, you are right it is not fully thread safe but it is thread safe for scenario that the call with deputy hostport won't be in the production peer list. E.g. a thread of request to deputy peer parking between line 244 and 245, a thread of request to production peer still has pre-seleced peers of c.selectedPeers[**hostport**], but this deputy peer won't be in production peer list for deputy scenario.

To be fully thread safe, a request go to production needs to use peer.begincall instead of subchannel.begincall since the later one would use peer score to decide which peer the request goes to. However, picking up a peer from subchannel.peerlist would violate load-balancing except there is only one peer(127.0.0.1:5437) in the peerlist.

Reply:
Add a RWLock for deputy request only. Please see comments inline